### ISSUE TITLE
Feature/472 create model abstraction for declearn integration - [merged]

### ISSUE THREAD

**Issue by srcansiz** [fedbiomed/fedbiomed#725]
In GitLab by @ybouilla on Feb 17, 2023, 17:30

_Merges feature/472-create-model-abstraction-for-declearn-integration -> develop_

**MR description**

Implements model abstraction for pytorch and scikit learn 
Closes #472.

Here a list of points that i found difficult to implement (and that need a closer look).

1. BaseSklearnModel:
- batch_size computation / reset that is done internally 
- sklearn model n_iter attribute increment/decrement (hidden through methods)
- apply_updates method: slightly modified from the one of the poc, that adds gradients instead of changing them, in the same spirit as the `apply_updates` of pytorch. Check if computation is correct
- Toolbox classes, that implement some method using multiple inheritance

2. TorchModel

- method train not implemented: should we compute loss in this method ? 
- is get_gradients method correct?
- saving state of the TorchModel: initial parameters are saved in a specific attribute: 
- make sure it handles frozen layers 

More broadly speaking:
 - appropriate raise of exceptions: some exceptions may not be caught
 - appropriate naming of variables
 - create multiple modules for each classes (ie one class per file) rather than having 3 classes in a file. 

**Developer Certificate Of Origin (DCO)**

By opening this merge request, you agree the
[Developer Certificate of Origin (DCO)](https://gitlab.inria.fr/fedbiomed/fedbiomed/-/blob/develop/CONTRIBUTING.md#fed-biomed-developer-certificate-of-origin-dco)

This DCO essentially means that:

- you offer the changes under the same license agreement as the project, and
- you have the right to do that,
- you did not steal somebody else’s work.

**License**

Project code files should begin with these comment lines to help trace their origin:
```
# This file is originally part of Fed-BioMed
# SPDX-License-Identifier: Apache-2.0
```

Code files can be reused from another project with a compatible non-contaminating license.
They shall retain the original license and copyright mentions.
The `CREDIT.md` file and `credit/` directory shall be completed and updated accordingly.

**Guidelines for MR review**

General:

* give a glance to [DoD](https://fedbiomed.gitlabpages.inria.fr/latest/developer/Fed-BioMed_DoD.pdf)
* check [coding rules and coding style](https://fedbiomed.gitlabpages.inria.fr/latest/developer/usage_and_tools/#coding-style)
* check docstrings (eg run `tests/docstrings/check_docstrings`)

Specific to some cases:

* update all conda envs consistently (`development` and `vpn`, Linux and MacOS)
* if modified researcher (eg new attributes in classes) check if breakpoint needs update (`breakpoint`/`load_breakpoint` in `Experiment()`, `save_state`/`load_state` in aggregators, strategies, secagg, etc.)
